##**Breast Cancer Detection using Logistic Regression & KNN Classification**

In [4]:
#importing libraries
import numpy as np
import sklearn.datasets

In [5]:
#getting the dataset
breast_cancer = sklearn.datasets.load_breast_cancer()

In [6]:
print(breast_cancer)
# 0 = malignant
# 1 = benign

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [7]:
X = breast_cancer.data
Y = breast_cancer.target


In [8]:
print(X)   #data
print(Y)   #equivalent labels

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

In [9]:
print(X.shape, Y.shape)

(569, 30) (569,)


##Importing data to Pandas Data Frame

In [10]:
import pandas as pd

In [11]:
data = pd.DataFrame(breast_cancer.data, columns=breast_cancer.feature_names)

In [12]:
data['class'] = breast_cancer.target

In [13]:
data.head(10)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   
5        12.45         15.70           82.57      477.1          0.12780   
6        18.25         19.98          119.60     1040.0          0.09463   
7        13.71         20.83           90.20      577.9          0.11890   
8        13.00         21.82           87.50      519.8          0.12730   
9        12.46         24.04           83.97      475.9          0.11860   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760         0.30010              0.14710         0.2419   
1           0.07864         0.08690              0.07017         0.1812   
2           0.15990         0.19740              0.12790         0.2069   
3           0.28390         0.24140              0.10520         0.2597   
4           0.13280         0.19800              0.10430         0.1809   
5           0.17000         0.15780              0.08089         0.2087   
6           0.10900         0.11270              0.07400         0.1794   
7           0.16450         0.09366              0.05985         0.2196   
8           0.19320         0.18590              0.09353         0.2350   
9           0.23960         0.22730              0.08543         0.2030   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   
5                 0.07613  ...          23.75           103.40       741.6   
6                 0.05742  ...          27.66           153.20      1606.0   
7                 0.07451  ...          28.14           110.60       897.0   
8                 0.07389  ...          30.73           106.20       739.3   
9                 0.08243  ...          40.68            97.65       711.4   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   
5            0.1791             0.5249           0.5355                0.1741   
6            0.1442             0.2576           0.3784                0.1932   
7            0.1654             0.3682           0.2678                0.1556   
8            0.1703             0.5401           0.5390                0.2060   
9            0.1853             1.0580           1.1050                0.2210   

   worst symmetry  worst fractal dimension  class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  
5          0.3985                  0.12440      0  
6          0.3063                  0.08368      0  
7          0.3196                  0.11510      0  
8          0.4378                  0.10720      0  
9          0.4366                  0.20750      0  

[10 rows x 31 co

In [14]:
data.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension       class  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [15]:
print(data['class'].value_counts())

1    357
0    212
Name: class, dtype: int64


In [16]:
print(breast_cancer.target_names)

['malignant' 'benign']


In [17]:
data.groupby('class').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

### 0 - malignant
### 1 - benign

##Train and Test Split

In [18]:
from sklearn.model_selection import train_test_split

In [19]:
x_train,x_test,y_train,y_test = train_test_split(X,Y)

In [20]:
print(Y.shape,y_train.shape,y_test.shape)

(569,) (426,) (143,)


In [21]:
x_train,x_test,y_train,y_test = train_test_split(X,Y,test_size=0.1) #Percentage of data held for testing i.e. 10% will be taken for testing

In [22]:
print(Y.shape,y_train.shape,y_test.shape)

(569,) (512,) (57,)


In [23]:
print(Y.mean(),y_train.mean(),y_test.mean())

0.6274165202108963 0.62890625 0.6140350877192983


In [24]:
x_train,x_test,y_train,y_test = train_test_split(X,Y,test_size=0.1,stratify=Y)
#stratify -> to overcome train_test_split's biased random sampling problem i.e overfitting and underfitting

In [25]:
print(Y.mean(),y_train.mean(),y_test.mean())

0.6274165202108963 0.626953125 0.631578947368421


In [26]:
x_train,x_test,y_train,y_test = train_test_split(X,Y,test_size=0.1,stratify=Y,random_state=1)
#It means one random_state value has a fixed dataset.
#It means every time we run code with random_state value 1, it will produce the same splitting datasets.

In [27]:
print(X.mean(),x_train.mean(),x_test.mean())

61.890712339519624 61.31637960106119 67.04963097269005


In [28]:
print(x_train)

[[1.490e+01 2.253e+01 1.021e+02 ... 2.475e-01 2.866e-01 1.155e-01]
 [1.205e+01 1.463e+01 7.804e+01 ... 6.548e-02 2.747e-01 8.301e-02]
 [1.311e+01 1.556e+01 8.721e+01 ... 1.986e-01 3.147e-01 1.405e-01]
 ...
 [1.258e+01 1.840e+01 7.983e+01 ... 8.772e-03 2.505e-01 6.431e-02]
 [1.349e+01 2.230e+01 8.691e+01 ... 1.282e-01 2.871e-01 6.917e-02]
 [1.919e+01 1.594e+01 1.263e+02 ... 1.777e-01 2.443e-01 6.251e-02]]


##**Logistic Regression**

In [29]:
from sklearn.linear_model import LogisticRegression

In [30]:
classifier = LogisticRegression()

In [31]:
#Training Model on Training Data
classifier.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

##Evaluation of model

In [32]:
from sklearn.metrics import accuracy_score

In [33]:
prediction_on_training_data = classifier.predict(x_train)
accuracy_train = accuracy_score(y_train,prediction_on_training_data)
print("Accuracy on training data = ", accuracy_train)

Accuracy on training data =  0.951171875


In [34]:
#prediction on Test data
prediction_on_testing_data = classifier.predict(x_test)
accuracy_test = accuracy_score(y_test,prediction_on_testing_data)
print("Accuracy on testing data = ", accuracy_test)

Accuracy on testing data =  0.9298245614035088


##Detecting Whether Patient Have Breast Cancer Is Malignant Or Benign


###*For Malignant*

In [35]:
input_data = (17.99,10.38,122.8,1001,0.1184,0.2776,0.3001,0.1471,0.2419,0.07871,1.095,0.9053,8.589,153.4,0.006399,0.04904,0.05373,0.01587,0.03003,0.006193,25.38,17.33,184.6,2019,0.1622,0.6656,0.7119,0.2654,0.4601,0.1189)

#Changing input_data to numpy array for prediction
inp_np_arr = np.asarray(input_data)
print(inp_np_arr)

#reshaping the array shape as we are predicting for one instance

in_re = inp_np_arr.reshape(1,-1)
print(in_re)

[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
 1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
 6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
 1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
 4.601e-01 1.189e-01]
[[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
  1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
  6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
  1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
  4.601e-01 1.189e-01]]


In [36]:
#prediction
stages = {0:"Malignant",1:"Benign"}
predict_op = classifier.predict(in_re)
print("Breast Cancer Is",stages[predict_op[0]])

Breast Cancer Is Malignant


###*For Benign*


In [37]:
input_data_b = (13.03,18.42,82.61,523.8,0.08983,0.03766,0.02562,0.02923,0.1467,0.05863,0.1839,2.342,1.17,14.16,0.004352,0.004899,0.01343,0.01164,0.02671,0.001777,13.3,22.81,84.46,545.9,0.09701,0.04619,0.04833,0.05013,0.1987,0.06169)

#Changing input_data to numpy array for prediction
inp_np_arr_b = np.asarray(input_data_b)
print(inp_np_arr_b)

#reshaping the array shape as we are predicting for one instance

in_re_b = inp_np_arr_b.reshape(1,-1)
print(in_re_b)

[1.303e+01 1.842e+01 8.261e+01 5.238e+02 8.983e-02 3.766e-02 2.562e-02
 2.923e-02 1.467e-01 5.863e-02 1.839e-01 2.342e+00 1.170e+00 1.416e+01
 4.352e-03 4.899e-03 1.343e-02 1.164e-02 2.671e-02 1.777e-03 1.330e+01
 2.281e+01 8.446e+01 5.459e+02 9.701e-02 4.619e-02 4.833e-02 5.013e-02
 1.987e-01 6.169e-02]
[[1.303e+01 1.842e+01 8.261e+01 5.238e+02 8.983e-02 3.766e-02 2.562e-02
  2.923e-02 1.467e-01 5.863e-02 1.839e-01 2.342e+00 1.170e+00 1.416e+01
  4.352e-03 4.899e-03 1.343e-02 1.164e-02 2.671e-02 1.777e-03 1.330e+01
  2.281e+01 8.446e+01 5.459e+02 9.701e-02 4.619e-02 4.833e-02 5.013e-02
  1.987e-01 6.169e-02]]


In [38]:
predict_op_b = classifier.predict(in_re_b)
print("Breast Cancer Is",stages[predict_op_b[0]])

Breast Cancer Is Benign


##**KNN Classification**

In [39]:
from sklearn.neighbors import KNeighborsClassifier

In [40]:
knn = KNeighborsClassifier(n_neighbors = 5)

In [41]:
knn.fit(x_train, y_train)

KNeighborsClassifier()

##Evaluation Of KNN Clasification Model

In [42]:
from sklearn.metrics import accuracy_score

In [43]:
#prediction on Training data
predict_on_training_data = knn.predict(x_train)
accuracy_train = accuracy_score(y_train,predict_on_training_data)
print("Accuracy on training data = ",accuracy_train)

Accuracy on training data =  0.94921875


In [55]:
#prediction on Test data
predict_test_data = knn.predict(x_test)
accuracy_test = accuracy_score(y_test,predict_test_data)
print("Accuracy on testing data = ",accuracy_test)
# OR
#print("Accuracy on testing data = ",knn.score(x_test,y_test))

Accuracy on testing data =  0.9298245614035088


##Detecting Whether Patient Have Breast Cancer Is Malignant Or Benign

###*For Malignant*

In [52]:
input_data = (17.99,10.38,122.8,1001,0.1184,0.2776,0.3001,0.1471,0.2419,0.07871,1.095,0.9053,8.589,153.4,0.006399,0.04904,0.05373,0.01587,0.03003,0.006193,25.38,17.33,184.6,2019,0.1622,0.6656,0.7119,0.2654,0.4601,0.1189)

#Changing input_data to numpy array for prediction
inp_np_arr1 = np.asarray(input_data)
print(inp_np_arr1)

#reshaping the array shape as we are predicting for one instance

in_re1 = inp_np_arr1.reshape(1,-1)
print(in_re1)

[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
 1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
 6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
 1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
 4.601e-01 1.189e-01]
[[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
  1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
  6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
  1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
  4.601e-01 1.189e-01]]


In [53]:
#prediction
stages = {0:"Malignant",1:"Benign"}
predict_op1 = knn.predict(in_re1)
print("Breast Cancer Is",stages[predict_op1[0]])

Breast Cancer Is Malignant


###*For Benign*

In [51]:
input_data_b1 = (13.03,18.42,82.61,523.8,0.08983,0.03766,0.02562,0.02923,0.1467,0.05863,0.1839,2.342,1.17,14.16,0.004352,0.004899,0.01343,0.01164,0.02671,0.001777,13.3,22.81,84.46,545.9,0.09701,0.04619,0.04833,0.05013,0.1987,0.06169)

#Changing input_data to numpy array for prediction
inp_np_arr_b1 = np.asarray(input_data_b1)
print(inp_np_arr_b1)

#reshaping the array shape as we are predicting for one instance

in_re_b1 = inp_np_arr_b1.reshape(1,-1)
print(in_re_b1)

[1.303e+01 1.842e+01 8.261e+01 5.238e+02 8.983e-02 3.766e-02 2.562e-02
 2.923e-02 1.467e-01 5.863e-02 1.839e-01 2.342e+00 1.170e+00 1.416e+01
 4.352e-03 4.899e-03 1.343e-02 1.164e-02 2.671e-02 1.777e-03 1.330e+01
 2.281e+01 8.446e+01 5.459e+02 9.701e-02 4.619e-02 4.833e-02 5.013e-02
 1.987e-01 6.169e-02]
[[1.303e+01 1.842e+01 8.261e+01 5.238e+02 8.983e-02 3.766e-02 2.562e-02
  2.923e-02 1.467e-01 5.863e-02 1.839e-01 2.342e+00 1.170e+00 1.416e+01
  4.352e-03 4.899e-03 1.343e-02 1.164e-02 2.671e-02 1.777e-03 1.330e+01
  2.281e+01 8.446e+01 5.459e+02 9.701e-02 4.619e-02 4.833e-02 5.013e-02
  1.987e-01 6.169e-02]]


In [54]:
#Prediction
stages = {0:"Malignant",1:"Benign"}
predict_op_b1 = knn.predict(in_re_b1)
print("Breast Cancer Is",stages[predict_op_b1[0]])

Breast Cancer Is Benign


## **Thus, overall accuracy of Logistic Regression is more than KNN Classification.**

***Accuracy (For Training Data):-***

Logistic Regression --> 95.11%

KNN Classification  --> 94.92%

***Accuracy (For Testing Data):-***

Logistic Regression --> 92.98%

KNN Classification  --> 92.98%